cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+  }
+  ASSERT_LE(diffs.size(), max_number_of_equal_diffs);
+
+  for (const auto& diff : diffs) {
+    uint64_t left_sum = 0, right_sum = 0;
+    for (const auto& val : diff.first) left_sum += val;
+    for (const auto& val : diff.second) right_sum += val;
+    ASSERT_EQ(left_sum, right_sum);
+    ASSERT_LE(diff.first.size(), 2);
+    ASSERT_LE(diff.second.size(), 2);
+  }
+
+  if (diffs.size() == 0) {
+    // no differences found, the arrays are equal
+    ASSERT_TRUE(original.page_lengths == modified.page_lengths);
+  }
+}
+
+uint64_t ElementCount(int64_t size, int32_t byte_width, bool nullable) {
+  if (nullable) {
+    // in case of nullable types the def_levels are also fed through the chunker
+    // to identify changes in the null bitmap, this will increase the byte width
+    // and decrease the number of elements per chunk
+    byte_width += 2;
+  }
+  return size / byte_width;
+}
+
+void AssertAllBetween(const ChunkList& chunks, int64_t min, int64_t max) {
+  // except the last chunk since it is not guaranteed to be within the range
+  for (size_t i = 0; i < chunks.size() - 1; i++) {
+    ASSERT_GE(chunks[i], min);
+    ASSERT_LE(chunks[i], max);
+  }
+  ASSERT_LE(chunks.back(), max);
+}
+
+void AssertChunkSizes(const std::shared_ptr<::arrow::DataType>& dtype,
+                      const RowGroupInfo& base_info, const RowGroupInfo& modified_info,
+                      bool nullable, bool enable_dictionary, int64_t min_chunk_size,
+                      int64_t max_chunk_size) {
+  if (dtype->id() != ::arrow::Type::BOOL) {
+    ASSERT_EQ(base_info.has_dictionary_page, enable_dictionary);
+    ASSERT_EQ(modified_info.has_dictionary_page, enable_dictionary);
+  }
+  if (::arrow::is_fixed_width(dtype->id())) {
+    // for nullable types we cannot calculate the exact number of elements because
+    // not all elements are fed through the chunker (null elements are skipped)
+    auto byte_width = (dtype->id() == ::arrow::Type::BOOL) ? 1 : dtype->byte_width();
+    auto min_length = ElementCount(min_chunk_size, byte_width, nullable);
+    auto max_length = ElementCount(max_chunk_size, byte_width, nullable);
+    AssertAllBetween(base_info.page_lengths, min_length, max_length);
+    AssertAllBetween(modified_info.page_lengths, min_length, max_length);
+  } else if (::arrow::is_base_binary_like(dtype->id()) && !enable_dictionary) {
+    AssertAllBetween(base_info.page_sizes, min_chunk_size, max_chunk_size);
+    AssertAllBetween(modified_info.page_sizes, min_chunk_size, max_chunk_size);
+  }
+}
+
+constexpr int64_t kMinChunkSize = 8 * 1024;
+constexpr int64_t kMaxChunkSize = 32 * 1024;
+constexpr int64_t kPartSize = 128 * 1024;
+constexpr int64_t kEditSize = 128;
+
+struct CaseConfig {
+  // Arrow data type to generate the testing data for
+  std::shared_ptr<::arrow::DataType> dtype;
+  // Whether the data type is nullable
+  bool is_nullable;
+  // Approximate number of bytes per record to calculate the number of elements to
```

Comment:
What is a "record" in this context? A top-level value?

Reply:
Yes. I removed this field from CaseConfig and moved the record size estimation to the test case `SetUp()`. The fixed width types can be explicitly calculated and using a constant for the variable sized types.

Currently this is only required to reduce the number of generated samples to avoid the tests heavily slowing down, added a comment about this.